GCTC-NTGC/gc-digital-talent
# Feature - 4029 - Theme Switcher

Resolves #4029 

> **Note**
> ~~If someone approves https://github.com/GCTC-NTGC/gc-digital-talent/pull/4131 first, I can see about using that component in this~~

## Summary

Introduces a new context and component to control the current colour theme (dark/light mode).

## TO DO

- [x] Merge `main` once #4049 is complete
- [x] Create story
- [x] Add dark mode styles
- [x] Fix background colour

## Testing

> **Note**
> This is only in storybook right now so all testing must take place there. Jest is testing that `localStorage.theme` is modified when buttons are clicked along with the `className` update which is derived from the `ThemeContext`.

1. Run the common storybook `npm run storybook --workspace=common`
2. Navigate to the Theme Switcher story
3. Click on the buttons and ensure the switcher updates
4. Bonus: Check to see if "theme" key exists in local storage with the appropriate value

## Additions

### Context: `<ThemeProvider>`

This is added within our existing `<ContextContainer>` just below the language container. This controls the current theme including manipulating the DOM.

### Hook: `useTheme`

This is a simple hook that allows us to interact with the context, retrieving the current theme mode along with a method to set it.

### Component: `<ThemeSwitcher>`

A new component (created by @JoshBeveridge ) has been created that uses the `useTheme` hook to update the current mode.

## Example

```tsx
const ExampleSwitcher = () => {
  const { mode, setMode } = useTheme();
  
  return (
     <div>
       Current Mode: {mode}
       <button onClick={() => setMode("pref"}>
         Toggle Preference
       </button>
       <button onClick={() => setMode("light"}>
         Toggle Light Mode
       </button>
       <button onClick={() => setMode("dark"}>
         Toggle Dark Mode
       </button>
    </div>
  ); 
}
```
